File: easy/0563_binary_tree_tilt/solution.cpp
#include "../../utils/cpp/nodes.hpp"

class Solution {
public:
  int findTilt(utils::TreeNode *root) {
    int tilt = 0;
    subtreeSum(root, tilt);
    return tilt;
  }

private:
  int subtreeSum(utils::TreeNode *node, int &tilt) {
    if (!node)
      return 0;

    int leftSum = subtreeSum(node->left, tilt);
    int rightSum = subtreeSum(node->right, tilt);

    tilt += abs(leftSum - rightSum);

    return node->val + leftSum + rightSum;
  }
};

File: easy/0563_binary_tree_tilt/Solution.java
import utils.java.nodes.TreeNode;

class Solution {
  private int tilt;

  public int findTilt(TreeNode root) {
    this.subtreeSum(root);
    return this.tilt;
  }

  private int subtreeSum(TreeNode node) {
    if (node == null) {
      return 0;
    }

    final int leftSum = this.subtreeSum(node.left);
    final int rightSum = this.subtreeSum(node.right);

    this.tilt += Math.abs(leftSum - rightSum);

    return node.val + leftSum + rightSum;
  }
}
